cms/tiny_table_block/utils.py

```diff
@@ -0,0 +1,70 @@
+from typing import TYPE_CHECKING
+
+from bs4 import BeautifulSoup
+
+if TYPE_CHECKING:
+    from bs4.element import Tag
+
+
+def get_cell_data(cell: "Tag") -> dict[str, str | int]:
+    cell_data = {"value": cell.text.strip(), "type": cell.name}
+
+    if (rowspan := int(cell.get("rowspan", 1))) > 1:
+        cell_data["rowspan"] = rowspan
+    if (colspan := int(cell.get("colspan", 1))) > 1:
+        cell_data["colspan"] = colspan
+    if scope := cell.get("scope"):
+        cell_data["scope"] = scope
+    if align := cell.get("align"):
+        cell_data["align"] = align
+
+    return cell_data
+
+
+def html_table_to_dict(content: str) -> dict:
+    """Take an HTML table and convert it to a dictionary.
+
+    The dictionary has the following structure:
+    - headers - a list of header row lists, each containing the cell info
+    - rows - a list of row lists, each containing the cell info
+    - html - the original html
+    """
+    # TODO: run content through nh3
+
+    soup = BeautifulSoup(content, "html.parser")
+
+    table = soup.find("table")
+    if not table:
+        return {
+            "headers": [],
+            "rows": [],
+            "html": "",
+        }
+
+    # Extract headers
+    headers = []
+    if thead := table.find("thead"):
+        for header_row in thead.find_all("tr"):
+            headers.append([get_cell_data(cell) for cell in header_row.find_all("th")])
+
+        if tbody_rows := table.find("tbody"):
+            table_rows = tbody_rows.find_all("tr")
+        else:
+            # If there's a thead but no tbody, get all rows not in thead
+            table_rows = [row for row in table.find_all("tr") if row.parent != thead]
+    else:
+        table_rows = table.find_all("tr")
+        # TODO: extract headers from tr if all cells are th?
```

Comment:
Can you clarify?. Is this needed given `get_cell_data` below is only looping over `td`/`th`?

Reply:
header rows are wrapped in `<thead>` by the design system - https://github.com/ONSdigital/design-system/blob/main/src/components/table/_macro.njk#L19

it is possible to paste the following table

```
<table>
<tr><th>header cell</th><th>header cell</th></tr>
<tr><td>cell</td><td>cell</td></tr>
</table>
```

which is valid.

I want to get this in front of folk sooner to get feedback, and sort this out while working on the other couple of table-related tickets (links and cell alignment)